Services/Init/classes/class.ilInitialisation.php

```diff
@@ -435,33 +439,66 @@ protected static function determineClient(): void
         if (!$DIC->isDependencyAvailable('iliasIni')) {
             self::abortAndDie('Fatal Error: ilInitialisation::determineClient called without initialisation of ILIAS ini file object.');
         }
-        $in_unit_tests = defined('IL_PHPUNIT_TEST');
-        $context_supports_persitent_session = ilContext::supportsPersistentSessions();
-        $can_set_cookie = !$in_unit_tests && $context_supports_persitent_session;
-        $has_request_client_id = $DIC->http()->wrapper()->query()->has('client_id');
-        $has_cookie_client_id = $DIC->http()->cookieJar()->has('ilClientId');
+
+        // determine the available clientIds (default, request, cookie)
         $default_client_id = $DIC->iliasIni()->readVariable('clients', 'default');
 
-        // determintaion of client_id:
+        if ($DIC->http()->wrapper()->query()->has('client_id')) {
+            $client_id_from_get = $DIC->http()->wrapper()->query()->retrieve(
+                'client_id',
+                // Refinery is not yet initialized during this state of init, requiring to provide a full implementation
+                // of the transformation.
+                new class () implements \ILIAS\Refinery\Transformation {
+                    /**
+                     * @inheritDoc
+                     */
+                    public function transform($from): string
+                    {
+                        if (!is_string($from)) {
+                            throw new InvalidArgumentException(__METHOD__ . " the argument is not a string.");
+                        }
+                        return strip_tags($from);
+                    }
+
+                    /**
+                     * @inheritDoc
+                     */
+                    public function applyTo(Result $result): Result
+                    {
+                        return $result->then(function ($value): Result {
+                            try {
+                                return new Ok($this->transform($value));
+                            } catch (Exception $exception) {
+                                return new Error($exception);
+                            }
+                        });
+                    }
+
+                    /**
+                     * @inheritDoc
+                     */
+                    public function __invoke($from)
+                    {
+                        return $this->transform($from);
+                    }
+                }
+            );
+        }
+        if ($DIC->http()->cookieJar()->has('ilClientId')) {
+            $client_id_from_cookie = $DIC->http()->cookieJar()->get('ilClientId')->getValue();
```

Comment:
Is it possible to use `$DIC->http()->wrapper()->cookie()->retrieve(...)` instead? Maybe you could re-use the anonymous class transformation from above.